libbpf-tools/filegone.c

```diff
@@ -0,0 +1,210 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */
+// filegone Trace why a file has vanished (either deleted or renamed).
+// Copyright 2024 Sony Group Corporation
+//
+// Based on filegone from BCC by Curu.
+//
+
+#include <argp.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <time.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "filegone.h"
+#include "filegone.skel.h"
+#include "btf_helpers.h"
+#include "trace_helpers.h"
+
+#define PERF_BUFFER_PAGES       16
+#define PERF_POLL_TIMEOUT_MS    100
+
+static volatile sig_atomic_t exiting = 0;
+
+static struct env {
+	pid_t pid;
+	bool verbose;
+} env = {};
+
+const char *argp_program_version = "filegone 0.1";
+const char *argp_program_bug_address =
+"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
+const char argp_program_doc[] =
+"Trace why a file has vanished (either deleted or renamed).\n"
+"\n"
+"USAGE: filegone  [--help] [-p PID]\n"
+"\n"
+"EXAMPLES:\n"
+"    filegone         # trace all events\n"
+"    filegone -p 123  # trace pid 123\n";
+
+static const struct argp_option opts[] = {
+	{"pid", 'p', "PID", 0, "Process PID to trace"},
+	{"verbose", 'v', NULL, 0, "Verbose debug output"},
+	{NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help"},
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	int pid;
+
+	switch (key) {
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'p':
+		errno = 0;
+		pid = strtol(arg, NULL, 10);
+		if (errno || pid <= 0) {
+			fprintf(stderr, "invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		env.pid = pid;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_int(int signo)
+{
+	exiting = 1;
+}
+
+const char *action2str(char action)
+{
+	return (action == 'D') ? "DELETE" : "RENAME";
+}
+
+void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	const char *action_str;
+	char file_str[96];
+	struct event e;
+	struct tm *tm;
+	char ts[32];
+	time_t t;
+
+	if (data_sz < sizeof(e)) {
+		printf("Error: packet too small\n");
+		return;
+	}
+
+	/* Copy data as alignment in the perf buffer isn't guaranteed. */
+	memcpy(&e, data, sizeof(e));
+	action_str = action2str(e.action);
+
+	if (strcmp(action_str, "RENAME") == 0) {
```

Comment:
we can classify the values of action using an enum type, without string comparison.